include/deal.II/base/tensor_function_parser.h

```diff
@@ -0,0 +1,361 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2005 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_tensor_function_parser_h
+#define dealii_tensor_function_parser_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/tensor.h>
+#include <deal.II/base/tensor_function.h>
+#include <deal.II/base/thread_local_storage.h>
+
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace mu
+{
+  class Parser;
+}
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declaration
+#ifndef DOXYGEN
+template <typename>
+class Vector;
+#endif
+
+/**
+ * This class implements a tensor function object that gets its value by parsing a
+ * string describing this function. It is a wrapper class for the muparser
+ * library (see http://muparser.beltoforion.de/). This class is essentially an extension
+ * of the FunctionParser class to read in a TensorFunction. The class reads in an
+ * expression of length dim<sup>rank<\sup> (separated by a semicolon) where the components
+ * of the tensor function are filled according to the C++ convention (fastest index is the
+ * most right one).
+ *
+ * @note In contrast to the FunctionParser class the TensorFunctionParser class does not support
+ * 	 automatic differentiation.
+ *
+ * A minimal example for the usage of the class would be:
+ * @code
+ * // set up time dependent tensor function:
+ * std::string variables = "x,y,t";
+ * std::string expression = "exp(-t)*cos(x+y);-sin(pi*x*y-t);sin(pi*x*y-t);exp(t)*cos(x+y)";
+ * std::map<std::string,double> constants;
+ * constants["pi"] = numbers::PI;
+ *
+ * // TensorFunctionParser with 2+1 variables (space + time) in 2D of rank 2:
+ * TensorFunctionParser<2,2> tfp;
+ * tfp.initialize(variables,
+ *               expression,
+ *               constants,
+ *               true);       // This tells the parser that
+ *                            // it is a time-dependent function
+ *                            // and there is another variable
+ *                            // to be taken into account (t).
```

Comment:
Could you move this comment up a line (i.e., right before the line with `true`) so that we don't have to split it across so many lines?